graylog2-web-interface/src/views/components/contexts/SearchPageLayoutContext.tsx

```diff
@@ -14,19 +14,21 @@
  * along with this program. If not, see
  * <http://www.mongodb.com/licensing/server-side-public-license>.
  */
+
 import * as React from 'react';
 
-type SearchPageLayoutConfig = {
-  sidebar: {
-    isPinned: boolean,
-  },
-};
+enum ViewActionsLayoutOptions {
+  FULL_MENU = 'FULL_MENU',
+  SAVE_COPY = 'SAVE_COPY',
+  BLANK = 'BLANK'
+}
 
-export type SearchPageLayout = {
-  config: SearchPageLayoutConfig,
-  actions: { toggleSidebarPinning: () => void },
-};
+export type LayoutState = {
+  sidebar: { isShown: boolean }
+  viewActionsLayoutOptions: ViewActionsLayoutOptions
```

Comment:
Just an idea, but maybe we can unify the structure of these settings, for example like this (they could be `true` by default): 
```
sidebar: { isShown: boolean },
dashboardActions: {
  save: { isShown: boolean },
  saveAs: { isShown: boolean },
  share: { isShown: boolean },
  actionsDropdown: { isShown: boolean },
},
```

This would make it easier to create a helper component for this logic, similar to `IfPermitted`. The usage could look like this: `<IfLayoutOption options=['dashboardActions.saveAs']>...</IfLayoutOption>`.

Another advantage, it would make it very clear what effect each setting has. One disadvantage is of course that we have to extend it over time and the state itself is larger compared with the current approach.

Alternatively we could use an enum for every option, in this case the usage of the helper component could look like this: `<IfLayoutOption options=['sidebar.COMPLETE']>...</IfLayoutOption>` and `<IfLayoutOption options=['dashboardActions.COMPLETE']>...</IfLayoutOption>`.

I think in every case there is a longterm benefit in unifying the structure of these settings.

Reply:
Just gonna jump in here and share my two cents too. I can definitely see the upsides to both of these approaches, and in my past experience I have been using the pattern that Linus is describing. However, I can see the points Mike is trying to make here, and I think it would be more advantageous for the component to decide what to render/ not render instead of a state describing what should be rendered. I think the example Mike is giving here is that in the future if another component were to consume that context and the fields we defined right now are not really relevant, it would be better if the component just takes in something like `FULL_MENU` and then decide how to interpret that in a way that's specific to that component. Otherwise we might have to expand the `ViewActionsLayoutOptions` and it could get really big.

Happy to hear what others think about this too.

Reply:
Thanks for everyone participating in this discussion, I appreciate the love for the detail and your thoughts about how to make this code sustainable. There are two potential sources of complexity here:

1. Having a complex state (by defining more fine-grained options on how and which elements are displayed) and simpler logic in the components
2. Having a simpler state (by selectining what is essentially a "preset") and more complex logic in the components (by testing which preset is selected and configuring elements according to it)

Both are potential sources for growth. For 1., the number of options could grow, for 2. it would be the number of presets and - in consequence - the logic in the components. I think there could be a middle way though, that leaves all of the pros of both approaches on the table and gets rid of some of the cons:

What we could do is define a larger state as proposed by @linuspahl, but define named default presets like this:

```
export const FULL_MENU = {
  save: { isShown: true },
  saveAs: { isShown: true },
  share: { isShown: true },
  actionsDropdown: { isShown: true },
};

export const SAVE_COPY = { ... };
```
for them which can be imported and used to have a consistent preset. This allows us to have simple logic in the components _and_ makes selecting/defaulting to named presets easy.

Btw: enums in TypeScript are rather clumsy and have a couple of disadvantages. I would always resort to plain constant maps and use union types instead.

Reply:
@dennisoelkers I think I have this way you requested if you could take a look and see if this is how you thought it would be implemented.

Reply:
@mikedklein: Yes, it looks good and is what had in mind. The only (minor) suggestion for improvement is that maybe `viewActions` could be better for the key, as it is shorter and - in the context of a `SearchPageLayoutContext` - eliminates a bit of redundancy.